pkg/image/qemu.go

```diff
@@ -273,3 +286,23 @@ func PreallocateBlankBlock(dest string, size resource.Quantity) error {
 
 	return nil
 }
+
+func addPreallocation(preallocate bool, args []string, fn func(args []string) ([]byte, error)) error {
+	var err error
+	preallocationMethod := 0
+	for retry := true; retry; retry = err != nil && preallocationMethod < maxPreallocationMethods {
+		var argsToTry []string
+		var output []byte
+		if preallocate {
+			klog.V(1).Info("Added preallocation")
+			argsToTry = append(args, preallocationMethods[preallocationMethod]...)
+		}
+		output, err = fn(argsToTry)
+		if err != nil && strings.Contains(string(output), "Unsupported preallocation mode") {
+			preallocationMethod++
+			klog.V(1).Infof("Unsupported preallocation mode. Retrying with %s", preallocationMethods[preallocationMethod])
```

Comment:
this code feels un-idiomatic. suggesting:
````
var err error
if !preallocate {
  return nil
}

for _, preallocationMethod := range preallocationMethods {
   klog.V(1).Infof("Added preallocation (method %s)", preallocationMethod)
   argsToTry = append(args, preallocationMethod)
   output, err = fn(argsToTry)
   if err == nil {
      break
   }
   if strings.Contains(string(output), "Unsupported preallocation mode") {
      klog.V(1).Infof("Unsupported preallocation mode %s", preallocationMethod)
   }
}

return err
````

Reply:
I feel like both variations are misbehaving with preallocate being false now that I think about it

Reply:
oh, it's a backport, I suppose we can leave it as-is and discuss improvements on main

Reply:
Hmm... You're right but the same code is in main. I'll create a new PR that fixes it upstream (in `main`), and we can later backport it.

What do you think?

Reply:
Sounds good!

Reply:
Here: https://github.com/kubevirt/containerized-data-importer/pull/1714